suite/tests/client-interface/drbbdup-emul-reg-clobber-test.dll.c

```diff
@@ -0,0 +1,311 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* Tests the drbbdup extension's interactions with emulation. */
+
+#include "dr_api.h"
+#include "client_tools.h"
+#include "drmgr.h"
+#include "drbbdup.h"
+#include "drutil.h"
+#include "drreg.h"
+#include <string.h>
+
+typedef struct _per_block_t {
+    bool saw_first;
+    bool saw_last;
+} per_block_t;
+
+/* Assume single threaded. */
+static uintptr_t encode_val = 1;
+
+#ifdef X86
+static bool saw_movs;
+#endif
+static bool has_rest_of_block_emul;
+
+static dr_emit_flags_t
+app2app_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+              bool translating)
+{
+#if VERBOSE
+    /* Useful for debugging. */
+    instrlist_disassemble(drcontext, tag, bb, STDERR);
+#endif
+
+    has_rest_of_block_emul = false;
+
+    /* Test drutil rep string expansion interacting with drbbdup. */
+    bool expanded;
+    if (!drutil_expand_rep_string_ex(drcontext, bb, &expanded, NULL))
+        DR_ASSERT(false);
+    if (expanded) {
+        /* We can't overlap ours with drutil's so bow out. */
+        has_rest_of_block_emul = true;
+        return DR_EMIT_DEFAULT;
+    }
+    /* Test handling of DR_EMULATE_REST_OF_BLOCK by inserting it in the middle
+     * of a multi-instr block and ensuring drbbdup doesn't let it leak through into
+     * the start of subsequent cloned blocks.
+     * We want to leave some blocks unchanged though to test drbbdup's own
+     * emulation labels.
+     *
+     * XXX i#5390: We should also test the last instr being emulated: not sure drbbdup
+     * will do the right thing there if it's a "special" instr.
+     */
+    static int bb_count;
+    if (++bb_count % 2 == 0)
+        return DR_EMIT_DEFAULT;
+    instr_t *mid = instrlist_last_app(bb);
+    if (mid == NULL || mid == instrlist_first_app(bb))
+        return DR_EMIT_DEFAULT;
+    mid = instr_get_prev_app(mid);
+    if (mid == NULL || mid == instrlist_first_app(bb))
+        return DR_EMIT_DEFAULT;
+    emulated_instr_t emulated_instr;
+    emulated_instr.size = sizeof(emulated_instr);
+    emulated_instr.pc = instr_get_app_pc(mid);
+    emulated_instr.instr = instr_clone(drcontext, mid);
+    emulated_instr.flags = DR_EMULATE_REST_OF_BLOCK;
+    drmgr_insert_emulation_start(drcontext, bb, mid, &emulated_instr);
+    /* XXX i#5400: We'd like to pass a user_data to the instrument event but drbbdup
+     * doesn't support that; instead we rely on being single-threaded.
+     */
+    has_rest_of_block_emul = true;
```

Comment:
All of this can be deleted (except the drutil_expand call above), right?

Reply:
Deleted.